libmamba/include/mamba/util/iterator.hpp

```diff
@@ -0,0 +1,400 @@
+// Copyright (c) 2019, QuantStack and Mamba Contributors
+//
+// Distributed under the terms of the BSD 3-Clause License.
+//
+// The full license is in the file LICENSE, distributed with this software.
+
+#ifndef MAMBA_UTIL_ITERATOR_HPP
+#define MAMBA_UTIL_ITERATOR_HPP
+
+#include <iterator>
+
+namespace mamba::util
+{
+    /**************************************
+     * class xbidirectional_iterator_base *
+     **************************************/
+
+    template <class I, class T, class D = std::ptrdiff_t, class P = T*, class R = T&>
+    class xbidirectional_iterator_base
+    {
+    public:
+
+        using derived_type = I;
+        using value_type = T;
+        using reference = R;
+        using pointer = P;
+        using difference_type = D;
+        using iterator_category = std::bidirectional_iterator_tag;
+
+        inline friend derived_type operator++(derived_type& d, int)
+        {
```

Comment:
`operator++()` has to be defined in the inheriting class, it is used for implementing other methods in this CRTP base class to avoid the same boilerplate code in every "final" iterator.